### ISSUE TITLE
Pyright infers wrong class types when __new__ method exists.

### ISSUE THREAD

**Issue by estheruary** [microsoft/pyright#3888]
**Describe the bug**
If you create a class with a `__new__` method and subclass it with a `__new__` method Pyright infers the wrong type.

**To Reproduce & Screenshots & Code**

```python
class FirstLevelMeta(type):
    def __new__(cls, name: str, bases, dct):
        new_class = super().__new__(cls, name, bases, dct)
        return new_class


class SecondLevelMeta(FirstLevelMeta):
    def __new__(cls, name: str, bases, dct):
        new_class = super().__new__(cls, name, bases, dct)

        return new_class


class ThirdLevel(SecondLevelMeta):
    def __new__(cls, name: str, bases, dct):
        new_class = super().__new__(cls, name, bases, dct)

        return new_class
```

Without `__new__`.

<img width="673" alt="image" src="https://user-images.githubusercontent.com/43628498/187780079-0775a36a-a8da-4d4c-ad81-caae4b4d0c1d.png">

With `__new__`

<img width="673" alt="image" src="https://user-images.githubusercontent.com/43628498/187780134-5e11cd8f-e93a-4ce2-b981-aad285728617.png">

**Expected behavior**
Pyright should recognize that the result of `__new__` is the new type being created rather than the type of the parent.

In `ThirdLevel` `reveal_type(new_class)` should be `Self@ThirdLevel` and not `Self@FirstLevel`.

**VS Code extension or command-line**
```json
    "pyright": {
      "version": "1.1.268",
      "resolved": "https://registry.npmjs.org/pyright/-/pyright-1.1.268.tgz",
      "integrity": "sha512-X+8N4BD4mCKFP5pjE5nGSHXz1ysIwCXS6t2Lnxz3kt85mOzh9Q2GNBNPYEW4svuRjfJm/8f7MIc4jyHFjsJwSw=="
    },
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

Type checkers need to handle `super()` calls in a special manner. In pyright, this involves a bunch of special-case code, some of which is a bit fragile. That's what led to this bug. Mypy also struggles with this as evidenced by the many bug reports involving super() calls. In any case, this particular problem should be gone now. I've added a unit test based on your code sample to prevent a future regression.

**Comment by estheruary** [comment]
Woo! 2/2 on stuff fixed in the next version! Thank you!

**Comment by erictraut** [comment]
We'll leave it open until the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.270, which I just published. It will also be included in a future release of pylance.
